### PR TITLE
Server side changes for sticky volumes

### PR DESCRIPTION
**Pull request by diptanu**

### CODE REVIEW

File: api/tasks.go [hunk 1]
```diff
@@ -121,6 +128,12 @@ func (g *TaskGroup) AddTask(t *Task) *TaskGroup {
 	return g
 }
 
+// RequireDisk adds a local disk to the task group
```
Comment (dadgar) [inline api/tasks.go]: Update the Copy() method of task group

File: jobspec/test-fixtures/basic.hcl [hunk 2]
```diff
@@ -42,6 +42,11 @@ job "binstore-storagelocker" {
       mode     = "delay"
     }
 
+    local_disk {
```
Comment (dadgar) [inline jobspec/test-fixtures/basic.hcl]: I would add a test that has two local_disk sections that will fail

File: command/validate_test.go [hunk 3]
```diff
@@ -32,7 +32,6 @@ job "job1" {
 			driver = "exec"
 			resources = {
 				cpu = 1000
-				disk = 150
```
Comment (dadgar) [inline command/validate_test.go]: Do you not have to put local disk?

File: api/tasks.go [hunk 4]
```diff
@@ -82,13 +82,20 @@ type Service struct {
 	Checks    []ServiceCheck
 }
 
+// LocalDisk is an ephemeral disk object
+type LocalDisk struct {
+	Sticky bool
+	DiskMB int `mapstructure:"disk"`
```
Comment (dadgar) [inline api/tasks.go]: Deprecate IOPS?

File: nomad/structs/structs.go [hunk 5]
```diff
@@ -2049,6 +2056,13 @@ func (t *Task) Validate() error {
 		mErr.Errors = append(mErr.Errors, err)
 	}
 
+	// Esnure the task isn't asking for disk resources
```
Comment (dadgar) [inline nomad/structs/structs.go]: Ensure

File: nomad/structs/structs.go [hunk 6]
```diff
@@ -1541,6 +1537,9 @@ type TaskGroup struct {
 	// Tasks are the collection of tasks that this task group needs to run
 	Tasks []*Task
 
+	// LocalDisk is the disk resources that the task group requests
+	LocalDisk *LocalDisk
```
Comment (dadgar) [inline nomad/structs/structs.go]: Add to the copy method

File: scheduler/generic_sched.go [hunk 7]
```diff
@@ -452,6 +452,7 @@ func (s *GenericScheduler) computePlacements(place []allocTuple) error {
 				TaskGroup:     missing.TaskGroup.Name,
 				Metrics:       s.ctx.Metrics(),
 				NodeID:        option.Node.ID,
+				Resources:     option.AllocResources,
```
Comment (dadgar) [inline scheduler/generic_sched.go]: You shouldn't have to do anything because the allocation has the task group name and you can look it up through the job

File: scheduler/rank.go [hunk 8]
```diff
@@ -44,6 +45,10 @@ func (r *RankedNode) SetTaskResources(task *structs.Task,
 	r.TaskResources[task.Name] = resource
 }
 
+func (r *RankedNode) SetAllocResources(resources *structs.Resources) {
```
Comment (dadgar) [inline scheduler/rank.go]: I think you can get rid of this

File: scheduler/generic_sched.go (hunk 7 above)
Comment (dadgar) [inline scheduler/generic_sched.go]: Add the resources in UpsertAllocs

**Comment by dadgar** [comment]
Add local disk to the diff

File: nomad/state/state_store.go [hunk 9]
```diff
@@ -357,6 +357,28 @@ func (s *StateStore) UpsertJob(index uint64, job *structs.Job) error {
...
+			continue
+		}
+		var diskMB int
+		for j, task := range tg.Tasks {
+			if task.Resources != nil {
+				resources := task.Resources
+				diskMB += resources.DiskMB
+				resources.DiskMB = 0
+				task.Resources = resources
+				tg.Tasks[j] = task
```
Comment (schmichael) [inline nomad/state/state_store.go]: This assignment is unnecessary. https://play.golang.org/p/doffmQ3xdd

File: nomad/state/state_store.go [hunk 10]
```diff
@@ -357,6 +357,28 @@ func (s *StateStore) UpsertJob(index uint64, job *structs.Job) error {
...
+			}
+		}
+		tg.LocalDisk = &structs.LocalDisk{
+			DiskMB: diskMB,
+		}
+		job.TaskGroups[i] = tg
```
Comment (schmichael) [inline nomad/state/state_store.go]: Same as above.

File: nomad/state/state_store.go [hunk 11]
```diff
@@ -357,6 +357,28 @@ func (s *StateStore) UpsertJob(index uint64, job *structs.Job) error {
 		return fmt.Errorf("unable to create job summary: %v", err)
 	}
 
+	// COMPAT 0.4.1 -> 0.5 Create the LocalDisk if it's nil by adding up DiskMB
+	// from task resources
+	for i, tg := range job.TaskGroups {
+		if tg.LocalDisk != nil {
+			continue
```
Comment (schmichael) [inline nomad/state/state_store.go]: Is it possible for a configuration to have both disk in Task Resources and Logical Disk, perhaps when transitioning from 0.4 -> 0.5? If so will this continue effectively ignore the 0.4 Resource config? Perhaps having both fails to validate?

File: nomad/state/state_store.go [hunk 12]
```diff
@@ -1690,6 +1712,29 @@ func (r *StateRestore) NodeRestore(node *structs.Node) error {
...
+				diskMB += resources.DiskMB
+				resources.DiskMB = 0
+				task.Resources = resources
+				tg.Tasks[j] = task
+			}
+		}
+		tg.LocalDisk = &structs.LocalDisk{
+			DiskMB: diskMB,
+		}
+		job.TaskGroups[i] = tg
```
Comment (schmichael) [inline nomad/state/state_store.go]: Same comments as above.

File: api/tasks.go (hunk 4 above)
Comment (diptanu) [inline api/tasks.go]: @dadgar Let's do this in another PR.

File: nomad/state/state_store.go (hunk 11 above)
Comment (diptanu) [inline nomad/state/state_store.go]: @schmichael No, since having disk resources in the resource block would result in a validation error. So once the user has migrated to a 0.5 setup the system shouldn't have any disk resource asks in the task resource config block.

File: command/validate_test.go (hunk 3 above)
Comment (diptanu) [inline command/validate_test.go]: @dadgar Not really, since a default local disk is always allocated.

File: jobspec/test-fixtures/basic.hcl (hunk 2 above)
Comment (diptanu) [inline jobspec/test-fixtures/basic.hcl]: Added

File: nomad/structs/structs.go (hunk 6 above)
Comment (diptanu) [inline nomad/structs/structs.go]: Good catch.

File: scheduler/context.go [hunk 13]
```diff
@@ -133,6 +133,13 @@ func (e *EvalContext) ProposedAllocs(nodeID string) ([]*structs.Allocation, erro
 	// Materialize the proposed slice
 	proposed = make([]*structs.Allocation, 0, len(proposedIDs))
 	for _, alloc := range proposedIDs {
+		if alloc.Job == nil {
```
Comment (dadgar) [inline scheduler/context.go]: I think we just introduce a `SharedResources *Resource` to the Allocation. Then we can store things that apply to all tasks there. And then you can add it to the `Resources` in the FSM and planner can use it to. Lets avoid copying and placing the job in the `ProposedAllocs`. Less performant and will be hard to remember the subtly of this, making it brittle

File: scheduler/generic_sched.go [hunk 14]
```diff
@@ -456,6 +456,9 @@ func (s *GenericScheduler) computePlacements(place []allocTuple) error {
 				DesiredStatus: structs.AllocDesiredStatusRun,
 				ClientStatus:  structs.AllocClientStatusPending,
 			}
+			alloc.SharedResources = &structs.Resources{
```
Comment (dadgar) [inline scheduler/generic_sched.go]: Just build inline above?

File: scheduler/system_sched.go [hunk 15]
```diff
@@ -320,6 +320,9 @@ func (s *SystemScheduler) computePlacements(place []allocTuple) error {
 				DesiredStatus: structs.AllocDesiredStatusRun,
 				ClientStatus:  structs.AllocClientStatusPending,
 			}
+			alloc.SharedResources = &structs.Resources{
```
Comment (dadgar) [inline scheduler/system_sched.go]: Here too

**Comment by dadgar** [comment]
LGTM
